kore/src/Kore/Step/SMT/Sorts.hs

```diff
@@ -0,0 +1,416 @@
+declareSortsWithNoJunkAxioms sorts =
+    SMT.declareDatatypes (mapMaybe buildSort sorts)
+  where
+    buildSort
+        ( attributes
+        , SentenceSort {sentenceSortParameters}
+        , SMTDataType { sort, constructors }
+        )
+      =
+        case (getSmtlib, getHook, sentenceSortParameters) of
+            (Nothing, Nothing, []) ->
+                case (maybeSortName, maybeSmtConstructors) of
+                    (Just sortName, Just smtConstructors) ->
+                        Just (sortName, [], smtConstructors)
+                    _ -> Nothing
+            (Nothing, Nothing, _ : _) ->
+                -- TODO(virgil): Handle parametric sorts.
+                Nothing
+            (_, _, _) -> (error . unlines)
+                [ "Unexpected smtlib or hook together with no-junk axiom:"
+                , "smtlib=" ++ show getSmtlib
+                , "hook=" ++ show getHook
+                , "sort=" ++ show sort
+                ]
+      where
+        maybeSortName = translateSortName sort
+        maybeSmtConstructors = mapM constructorToSmt constructors
+
+        Smtlib {getSmtlib} = Attribute.smtlib attributes
+        Hook {getHook} = Attribute.hook attributes
+
+    constructorToSmt
+        :: SMTConstructor -> Maybe (Text, [(Text, SMT.SExpr)])
+    constructorToSmt
+        SMTConstructor { name, argumentSorts }
+      = do
+        args <- zipWithM (constructorArgToSMT encodedName) [1..] argumentSorts
+        return (encodedName, args)
+      where
+        encodedName = encodeName name
+
+    constructorArgToSMT
+        :: Text -> Int -> Sort Object -> Maybe (Text, SMT.SExpr)
+    constructorArgToSMT constrName index constrSort = do
+        translatedSort <- translateSort constrSort
+        return (constrName <> pack (show index), translatedSort)
+
+parseNoJunkAxioms
+    :: IndexedModule
+        param
+        (KorePattern
+            Domain.Builtin
+            Variable
+            (Unified (Valid (Unified Variable)))
+        )
+        Symbol
+        Attribute.Axiom
+    -> [SMTDataType]
+parseNoJunkAxioms indexedModule =
+    filter (\ SMTDataType{constructors} -> not (null constructors))
+    $ mapMaybe
+        parseSMTDataType
+        (recursiveIndexedModuleAxioms indexedModule)
+
+parseSMTDataType
+    ::  ( Attribute.Axiom
+        , SentenceAxiom param VerifiedKorePattern
+        )
+    -> Maybe SMTDataType
+parseSMTDataType (attributes, SentenceAxiom {sentenceAxiomPattern})
+  | Axiom.Constructor.isConstructor (Attribute.constructor attributes)
+  = case fromKorePattern Object sentenceAxiomPattern of
+    Left err -> (error . unlines)
+        [ "Unexpected error transforming kore pattern to pure pattern:"
+        , " all patterns should be pure."
+        , "err=" ++ show err
+        , "pattern=" ++ show sentenceAxiomPattern
+        ]
+    Right patt -> parseSMTDataTypePattern patt
+  | otherwise = Nothing
+
+parseSMTDataTypePattern :: CommonStepPattern Object -> Maybe SMTDataType
```

Comment:
Is this pattern always supposed to come from a no-junk axiom?

Reply:
Yes.